Benchmarking recipes (Lauer et al.)

axel-lauer opened the pull request:
## Description

This PR implements a set of benchmarking recipes for comparison of different metrics (RMSE, bias, correlation, EMD) calculated for a given model simulation to the results from an ensemble of (model) datasets:
- [model_evaluation/recipe_model_benchmarking_annual_cycle.yml](https://github.com/ESMValGroup/ESMValTool/blob/benchmarking_maps4monitoring/esmvaltool/recipes/model_evaluation/recipe_model_benchmarking_annual_cycle.yml) (annual cycle)
- [model_evaluation/recipe_model_benchmarking_boxplots.yml](https://github.com/ESMValGroup/ESMValTool/blob/benchmarking_maps4monitoring/esmvaltool/recipes/model_evaluation/recipe_model_benchmarking_boxplots.yml) (boxplots)
- [model_evaluation/recipe_model_benchmarking_diurnal_cycle.yml](https://github.com/ESMValGroup/ESMValTool/blob/benchmarking_maps4monitoring/esmvaltool/recipes/model_evaluation/recipe_model_benchmarking_diurnal_cycle.yml) (diurnal cycle)
- [model_evaluation/recipe_model_benchmarking_maps.yml](https://github.com/ESMValGroup/ESMValTool/blob/benchmarking_maps4monitoring/esmvaltool/recipes/model_evaluation/recipe_model_benchmarking_maps.yml) (map plots of 2-dim variables)
- [model_evaluation/recipe_model_benchmarking_timeseries.yml](https://github.com/ESMValGroup/ESMValTool/blob/benchmarking_maps4monitoring/esmvaltool/recipes/model_evaluation/recipe_model_benchmarking_timeseries.yml) (time series)
- [model_evaluation/recipe_model_benchmarking_zonal.yml](https://github.com/ESMValGroup/ESMValTool/blob/benchmarking_maps4monitoring/esmvaltool/recipes/model_evaluation/recipe_model_benchmarking_zonal.yml) (zonal mean plots of 3-dim variables)

For this, the existing monitoring diagnostics [monitoring/monitor.py](https://github.com/ESMValGroup/ESMValTool/compare/main...benchmarking_maps4monitoring#diff-b95b5707ead9897ec9ee1a461b8e8adb6fe1b9a38bd8e4c64dd98efda62300b7) and [monitor/multi_datasets.py](https://github.com/ESMValGroup/ESMValTool/compare/main...benchmarking_maps4monitoring#diff-f21e88267deafc39c7c43a583ec36ea14f5ef7de5e4d140ec06616d55f838a52) have been extended.

The new diurnal cycle plot has also been added to the following existing recipes:
- [model_evaluation/recipe_model_evaluation_clouds_cycles.yml](https://github.com/ESMValGroup/ESMValTool/compare/main...benchmarking_maps4monitoring#diff-63833e02124aedd1b9add0cc227fab503ef7aef1cf28b867970fd4c967d7539b)
- [monitor/recipe_monitor.yml](https://github.com/ESMValGroup/ESMValTool/compare/main...benchmarking_maps4monitoring#diff-43059c33ac12104336414da08a54d72ddc25a4415ca62a1551328354659d07ca)
- [monitor/recipe_monitor_with_refs.yml](https://github.com/ESMValGroup/ESMValTool/compare/main...benchmarking_maps4monitoring#diff-1aba6aa2eb52e071dc129fcbbb7bab07fb2ff53304b2c6e0316b1b68d4b43ca4)

Documentation for the benchmarking recipes is available in [recipes/recipe_benchmarking.rst](), the documentation for [monitoring](https://github.com/ESMValGroup/ESMValTool/compare/main...benchmarking_maps4monitoring#diff-2853b01a87145be3509c4e75aa4498d4bc11501cb8b14c93cc30671557a7c502) and [model evaluation](https://github.com/ESMValGroup/ESMValTool/compare/main...benchmarking_maps4monitoring#diff-7cd2e4f80bfdbca43c77463b04f136588183f57516d2669f46cfb32b8a2fe76e) have been updated to include the diurnal cycle plots.

Closes #3498 

## Note for testing
The benchmarking recipes require the new preprocessor functions [local_solar_time](https://github.com/ESMValGroup/ESMValCore/pull/2258) and [distance_metric](https://github.com/ESMValGroup/ESMValCore/pull/2299) and the extended version of preprocessor [resample_hours](https://github.com/ESMValGroup/ESMValCore/pull/2440).
The recipe creating the portrait diagram additionally requires PR https://github.com/ESMValGroup/ESMValTool/pull/3551 providing the diagnostic script for this plot.

## Checklist

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

<!-- The next two lines turn the 🛠 and 🧪 below into hyperlinks -->
[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [ ] [🛠][1] This pull request has a [descriptive title](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#pull-request-title)
- [ ] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#code-quality)
- [ ] [🛠][1] [Documentation](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#documentation) is available
- [ ] [🛠][1] [Tests](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#tests) run successfully
- [ ] [🛠][1] The [list of authors](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#list-of-authors) is up to date
- [ ] [🛠][1] Any changed dependencies have been [added or removed correctly](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#dependencies)
- [ ] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#pull-request-checks) were successful

### [New or updated recipe/diagnostic](https://docs.esmvaltool.org/en/latest/community/diagnostic.html)

- [ ] [🧪][2] [Recipe runs successfully](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#testing-recipes)
- [ ] [🧪][2] [Recipe is well documented](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#recipe-and-diagnostic-documentation)
- [ ] [🧪][2] [Figure(s) and data](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#diagnostic-output) look as expected from literature
- [ ] [🛠][1] [Provenance information](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#recording-provenance) has been added

alistairsellar commented on esmvaltool/diag_scripts/monitor/multi_datasets.py:
```diff
@@ -1862,6 +2209,140 @@ def _get_reference_dataset(self, datasets):
+        """Get dataset to be benchmarked."""
+        variable = datasets[0][self.cfg['group_variables_by']]
+        benchmark_datasets = [d for d in datasets if
+                              d.get('benchmark_dataset', False)]
```
Should this really be False? It looks like it's selecting datasets that specify `benchmark_dataset: false`. If correct, please add a comment to avoid confusion.

axel-lauer replied:
As far as I understand, the Python function `dictionary.get(keyname, value)` returns `value` if the specified key does not exist, so setting `value` to `False` is just to make sure that datasets which do not have a 'benchmark_dataset' key are not selected.

alistairsellar commented on esmvaltool/diag_scripts/monitor/multi_datasets.py:
```diff
@@ -1862,6 +2209,140 @@ def _get_reference_dataset(self, datasets):
+                f"'{variable}', got {len(ref_datasets)}")
+        return None
+
+    def _get_benchmark_datasets(self, datasets):
```
Is there documentation (in code or docs) to indicate that user should specify `benchmark_dataset: true`  in recipe, to select that dataset to be benchmarked? This is clear from example recipes, but should really be captured somewhere beyond those.

Worth noting in that documentation that some plots will (I think) only plot a single benchmark even if multiple specified.

axel-lauer replied:
Good point. I'll add that to the documentation.

alistairsellar commented on esmvaltool/recipes/model_evaluation/recipe_model_benchmarking_annual_cycle.yml:
```diff
@@ -0,0 +1,197 @@
+          annual_cycle:
+            annual_mean_kwargs: False
+            plot_kwargs:
+              'MIROC6':
```
Does the choice of model here need to be consistent with selection of `benchmark_dataset` above?

axel-lauer replied:
It would make sense to be consistent if the benchmark_dataset is to be plotted with a special line color and style. I'll add this to the documentation.

alistairsellar reviewed (changes requested):
Great work @schlunma and @axel-lauer !

Some minor suggestions and questions from me.

I note also that readthedocs build still failing (so I haven't fully reviewed the documentation), and a couple of style complaints in run_tests.

axel-lauer commented on esmvaltool/diag_scripts/monitor/monitor.py:
```diff
@@ -27,6 +27,11 @@
       coordinates of length > 1. Supported coordinates: `time`, `shape_id`
       (optional) and `region` (optional).
+    - Diurnal cycle (plot type ``diurnal_cycle``): Generate a diurnal cycle
+      plot (timeseries like climatological from 0 to 24 hours). It will
```
That is a data limitation. We wrote 0-24 hours as (if data are available), one could plot 0:00:00 to 23:59:59.

alistairsellar replied:
> That is a data limitation. 
Does MIROC not provide a full hourly timeseries? Or some days are missing some hours and the diurnal cycle mean requires them all?

> We wrote 0-24 hours as (if data are available), one could plot 0:00:00 to 23:59:59.
Thanks, that makes sense.

axel-lauer replied:
The 3-hourly CMIP6 output used in the example provides data at 1.30 (0.00-3.00), 4.30 (3.00-6.00), ..., 22.30 (21.00-24.00). CMIP6 data are converted to full hours by preprocessor `resample_hours` so that they can be compared to ERA5 data (provided at full hours). The preprocessor `distance_metric,` applied afterwards requires all datasets to have the same time coordinate.

axel-lauer commented on esmvaltool/diag_scripts/monitor/monitor.py:
```diff
@@ -280,6 +290,57 @@ def plot_annual_cycle(self, cube, var_info):
+        The hourly climatology is done inside the function so the users can
+        plot both the timeseries and the diurnal cycle in one go
+        """
+        if 'diurnal_cycle' not in self.plots:
```
That is right, I just followed the structure that was there already.

axel-lauer commented on esmvaltool/diag_scripts/monitor/multi_datasets.py:
```diff
@@ -1779,6 +1929,195 @@ def _plot_hovmoeller_time_vs_lat_or_lon_without_ref(self, plot_func,
+        return (plot_path, {netcdf_path: cube})
+
+    def _plot_benchmarking_boxplot(self, df, cubes, variables, datasets):
+        """Plot benchmarking boxplot."""
```
@LisaBock Would you be able to address this comment?

LisaBock replied:
Thanks @alistairsellar! I just added a link to the recipe description. When using the recipe this information might be really useful.

alistairsellar replied:
Perfect - thanks @LisaBock!

bettina-gier reviewed (changes requested):
Look good in general, just a few comments:

- Lukas changed his name to Lindenlaub, so replace all the Ruhe's by Lindenlaub, other than in the paper citation
- There's a bunch of commented out code. If it's unnecessary, it should be removed before merging
- Diurnal_cycle recipe needs adjustment in the interpolate option of the resample_hours preprocessor, "True" is not a valid option with the current development Core.

Otherwise, the provenance works, the plots are consistent with the ones you put in the documentation and comparable to those in the paper, so just a few changes and you can get my approval. Fixing the interpolate option for the recipe should also fix the CircleCI error.

schlunma commented on esmvaltool/diag_scripts/monitor/multi_datasets.py:
```diff
@@ -65,6 +68,14 @@
       :func:`esmvalcore.preprocessor.regrid` for this). Input data
       needs to be 2D with dimensions `time`, `latitude`/`longitude`.
 
+    Benchmarking plots
```
I think it would be nice to to briefly explain these plot types (like it's done for the other ones). Also, this list is currently not rendered properly: https://esmvaltool--3598.org.readthedocs.build/en/3598/api/esmvaltool.diag_scripts.monitor/multi_datasets.html

schlunma commented on esmvaltool/diag_scripts/monitor/multi_datasets.py:
```diff
@@ -300,6 +311,7 @@
     Fontsize used for ticks, labels and titles. For the latter, use the given
     fontsize plus 2. Does not affect suptitles.
 log_y: bool, optional (default: True)
+    variables = var_order
```
```suggestion
```

I guess this shouldn't be here?

schlunma commented on esmvaltool/diag_scripts/monitor/multi_datasets.py:
```diff
@@ -589,6 +601,24 @@
     lat/lon on x-axis.
 
 
+Configuration options for plot type ``benchmarking_annual_cycle``
```
From the default options that are set below, it looks like there are indeed some configuration options for these plot types, e.g., `gridline_kwargs` for `benchmarking_annual_cycle`. It would be great to list them here or at least mention that they are the same (I guess?) as for non-benchmark version.

axel-lauer replied:
Added description for all benchmarking plots.

schlunma commented on esmvaltool/diag_scripts/monitor/multi_datasets.py:
```diff
@@ -651,9 +682,14 @@ def __init__(self, config):
 
         # Get default settings
         self.cfg = deepcopy(self.cfg)
+        self.cfg.setdefault('add_ancillary_variables', False)
+        self.cfg.setdefault('add_aux_coords', False)
+        self.cfg.setdefault('add_cell_measures', False)
         self.cfg.setdefault('facet_used_for_labels', 'dataset')
+        self.cfg.setdefault('facets_as_columns', [])
         self.cfg.setdefault('figure_kwargs', {'constrained_layout': True})
         self.cfg.setdefault('group_variables_by', 'short_name')
+        self.cfg.setdefault('groupby_facet', 'short_name')
```
It looks like these newly added options are not used. I guess they have been accidentally copied over from the seaborn diagnostic?

axel-lauer replied:
Not sure where those came from. Deleted the 4 unused entries.

schlunma commented on doc/sphinx/source/recipes/recipe_benchmarking.rst:
```diff
@@ -0,0 +1,140 @@
+Overview
+--------
+
+These recipes and diagnostics are based on :ref:`recipe_monitor <recipe_monitor>`: that allow plotting arbitrary preprocessor output, i.e., arbitrary variables from arbitrary datasets. An extension of these diagnostics is used to benchmark a model simulation with other datasets (e.g. CMIP6). The benchmarking features are described in `Lauer et al.`_:.
```
```suggestion
These recipes and diagnostics are based on :ref:`recipe_monitor <recipe_monitor>`: that allow plotting arbitrary preprocessor output, i.e., arbitrary variables from arbitrary datasets. An extension of these diagnostics is used to benchmark a model simulation with other datasets (e.g. CMIP6). The benchmarking features are described in `Lauer et al.`_.
```

schlunma commented on doc/sphinx/source/recipes/recipe_benchmarking.rst:
```diff
@@ -0,0 +1,140 @@
+.. code-block:: yaml
+
+   scripts:
+         allplots:
```
Indentation is messed up.

schlunma commented on doc/sphinx/source/recipes/recipe_monitor.rst:
```diff
@@ -189,6 +189,18 @@ Timeseries of tas including a reference dataset.
+.. _fig_diurnal_cycle:
+.. figure::  /recipes/figures/monitor/diurnalcycle_pr_tropics_EC-Earth3_3hr_historical_r1i1p1f1.png
+   :align:   center
+   :width:   14cm
```
Please add a figure caption

axel-lauer replied:
Done. I also indented all figure captions so they show up correctly in the pdf.

schlunma reviewed (commented):
Thanks Axel, looks great already! I have some additional comments on the docs (won't look into more detail as @bettina-gier already covered that, thanks!!).

axel-lauer commented on esmvaltool/recipes/model_evaluation/recipe_model_benchmarking_diurnal_cycle.yml:
```diff
@@ -0,0 +1,203 @@
+    resample_hours:
+      interval: 3
+      offset: 1
+      interpolate: true
```
Good catch, thank you!

axel-lauer commented on esmvaltool/references/lauer24gmd.bibtex:
```diff
@@ -0,0 +1,13 @@
+@article{lauer24gmd,
+	doi = {},
+	url = {},
+	year = ,
+	publisher = {European Geosciences Union},
+        address = {Göttingen Germany},
+	volume = {},
+	number = {},
+	pages = {},
+	author = {Axel Lauer and Lisa Bock and Birgit Hassler and Patrick Jöckel and Lukas Ruhe and Manuel Schlund},
+	title = {Monitoring and benchmarking Earth System Model simulations with ESMValTool v2.12.0},
+	journal = {Geoscientific Model Development}
+}
```
Waiting for the paper to be published. Should be very soon since it has been accepted for publication end of 2024.

axel-lauer replied:
I am now refering to the preprint as I don't know how long it takes until the final paper is out. I'll change this to the final version of the paper later or in a separate PR.

axel-lauer commented on doc/sphinx/source/recipes/recipe_benchmarking.rst:
```diff
@@ -0,0 +1,140 @@
+References
+----------
+
+* Lauer, A., L. Bock, B. Hassler, P. Jöckel, L. Ruhe, and M. Schlund: Monitoring and benchmarking Earth System Model simulations with ESMValTool v2.12.0, Geosci. Model Dev., xx, xxxx-xxxx,
```
Thanks! I think the paper should be available very soon (paper has been accepted for publication end of 2024) and I'll update the information then.

axel-lauer replied:
I am now refering to the preprint as I don't know how long it takes until the final paper is out. I'll change this to the final version of the paper later or in a separate PR.

axel-lauer commented on esmvaltool/recipes/model_evaluation/recipe_model_benchmarking_boxplots.yml:
```diff
@@ -0,0 +1,440 @@
+    - lauer_axel
+    - bock_lisa
+    - hassler_birgit
+    - ruhe_lukas
```
Corrected.

axel-lauer commented on esmvaltool/recipes/model_evaluation/recipe_model_benchmarking_timeseries.yml:
```diff
@@ -0,0 +1,208 @@
+    - lauer_axel
+    - bock_lisa
+    - hassler_birgit
+    - ruhe_lukas
```
Corrected.

axel-lauer commented on esmvaltool/recipes/model_evaluation/recipe_model_benchmarking_maps.yml:
```diff
@@ -0,0 +1,117 @@
+    - lauer_axel
+    - bock_lisa
+    - hassler_birgit
+    - ruhe_lukas
```
Done.

axel-lauer commented on doc/sphinx/source/recipes/recipe_benchmarking.rst:
```diff
@@ -0,0 +1,140 @@
+   :align:   center
+   :width:   10cm
+
+5-year annual mean bias (2000-2004) of the zonally averaged temperature in K from a historical simulation of MIROC6 compared with ERA5 reanalysis data. The stippled areas mask grid cells where the absolute BIAS (${\abs{BIAS}}$) is smaller than the maximum of the absolute 10% (${\abs{p10}}$) and the absolute 90% (${\abs{p90}}$) percentiles from an ensemble of CMIP6 models, i.e. ${\abs{BIAS} \geq m ...
```
Corrected, now showing up correctly in the html and pdf. Thanks for spotting this!

axel-lauer commented:
Thank you @bettina-gier for your review!

>     * Lukas changed his name to Lindenlaub, so replace all the Ruhe's by Lindenlaub, other than in the paper citation
Corrected. Thanks for spotting this!

>     * There's a bunch of commented out code. If it's unnecessary, it should be removed before merging
Removed commented out code from multi_datasets.py.
 
>     * Diurnal_cycle recipe needs adjustment in the interpolate option of the resample_hours preprocessor, "True" is not a valid option with the current development Core.
Also corrected, I guess I forgot to adjust this when the preprocessor options changed. Thanks!

axel-lauer commented on esmvaltool/diag_scripts/monitor/multi_datasets.py:
```diff
@@ -1779,6 +1929,195 @@ def _plot_hovmoeller_time_vs_lat_or_lon_without_ref(self, plot_func,
+                    " for '%s'",
+                    self._get_label(dataset))
+
+        # Make sure that the data has the correct dimensions
+        cube = dataset['cube']
```
Thanks for pointing this out. You are correct about the dimension check and also about the missing documentation on the expected data shape. To make things clearer, I added documentation to the code that the functions expects the same data (including the shape) as the plot type `zonal_mean_profile` for which @schlunma documented everything very nicely.

bettina-gier reviewed (approved):
I'll approve it now, thanks for addressing the comments!
Do you want to move the recipe for the new perfmetrics into that PR so it gets merged together and this PR doesn't depend on it anymore?

schlunma reviewed (commented):
I just pushed some commits to make this PR up-to-date with other developments of the monitoring diagnostic that have been merged while this PR was open. The merge with main only adapted existing lines, but not those added via this PR. I also fixed some minor issues with the doc about allowed options for these diagnostics.

I also fixed a bug which lead to no output from the boxplot diagnostic. Now all recipes run fine and produce the expected output (just tested) 🎉 

The only remaining comment (apart from two minor ones on the code) I have is about the `lauer25gmd` recipes. These include data from the EMAC model, which is not easily available. Usually we don't include these kind of recipes in the main repository because they are not really reproducible without the data. The alternative would be to put them on Zenodo (e.g., https://zenodo.org/records/7254313).

I know that the paper is almost published and it's probably hard to make any additional changes to it, but if that's possible, putting the recipe on Zenodo would be the preferred solution. If that's not possible, it would be great to add a note about that (ideally with a link to the data). We also need to make sure that those recipes are not part of the automated recipe test workflow.

axel-lauer commented on doc/sphinx/source/recipes/recipe_benchmarking.rst:
```diff
@@ -0,0 +1,175 @@
+
+These recipes and diagnostics are based on :ref:`recipe_monitor <recipe_monitor>`: that allow plotting arbitrary preprocessor output, i.e., arbitrary variables from arbitrary datasets. An extension of these diagnostics is used to benchmark a model simulation with other datasets (e.g. CMIP6). The benchmarking features are described in `Lauer et al.`_.
+
+.. _`Lauer et al.`: A. Lauer, Bock, L., Hassler, B., Jöckel, P., Ruhe, L., and Schlund, M.: Monitoring and benchmarking Earth System Model simulations with ESMValTool v2.12.0, Geosci. Model Dev., 2025.
```
I put in a link to the preprint as I don't know how long it takes until the final paper is out. I'll change this to the final version of the paper later or in a separate PR.

axel-lauer commented:
> The only remaining comment (apart from two minor ones on the code) I have is about the `lauer25gmd` recipes. These include data from the EMAC model, which is not easily available. Usually we don't include these kind of recipes in the main repository because they are not really reproducible without the data. The alternative would be to put them on Zenodo (e.g., https://zenodo.org/records/7254313).

Very good point! I removed the 'lauer25gmd' recipes from the branch and put them on Zenodo. The recipes are available at [10.5281/zenodo.11198444](https://doi.org/10.5281/zenodo.11198444), this will be updated in the paper once we get the proofs.

alistairsellar reviewed (approved):
Thanks for the great work everyone. I'm happy that my comments have been addressed (thanks!) so I'll approve now.

A note for future reviews - it really helps the reviewer if when you address a comment with e.g. "I have changed x" or "I have added this to y", your message also includes a link to the commit that addresses the comment. It makes it quicker for the reviewer to check what has changed in response.

schlunma commented on doc/sphinx/source/recipes/recipe_benchmarking.rst:
```diff
@@ -0,0 +1,175 @@
+* recipe_model_benchmarking_diurnal_cycle.yml
+* recipe_model_benchmarking_maps.yml
+* recipe_model_benchmarking_timeseries.yml
+* recipe_model_benchmarking_zonal.yml
```
Add portrait plot recipe?

schlunma commented on esmvaltool/recipes/model_evaluation/recipe_model_evaluation_portraits.yml:
```diff
@@ -0,0 +1,535 @@
+          label: Relative RMSE
+          extend: both
+
+  # perfmetrics_emd:  # seasonal normalized emd
```
Please remove commented parts

axel-lauer replied:
I removed the portrait plot recipe to avoid duplication with the recently merged recipe `recipe_portrait_CMIP.yml`.

schlunma reviewed (commented):
Thanks Axel for addressing all my comments and sorry for pushing while you were working on it, I think my changes broke the tests, so I wanted to fix them quickly.

I have two tiny comments, then will approve! Tests are failing because this requires a new ESMValCore release. We can merge once the first release candidate is out.

Cheers! 🚀

axel-lauer commented:
> I have two tiny comments, then will approve! Tests are failing because this requires a new ESMValCore release. We can merge once the first release candidate is out.

Should be all done now!! Thanks for your review and help with this PR!

schlunma reviewed (approved):
Awesome, thanks @axel-lauer and everyone!

sloosvel commented:
Hi, is this ready to be merged? The code freeze for v2.12.0 will be on Monday, make sure to have the PR merged by then if you want it included in the release!

axel-lauer commented:
Apart from the failing tests, this is ready to be merged. It is quite important that this goes into v2.12.0 as we promised this in our latest publication. I would guess the tests will be OK once the new Core release gets picked up by the test framework. Not sure how long this takes, though.

axel-lauer commented:
Yay! Thanks @schlunma !!!
